api/src/main/java/marquez/db/DatasetFacetsDao.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2018-2022 contributors to the Marquez project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package marquez.db;
+
+import static marquez.db.Columns.toPgObject;
+
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import lombok.NonNull;
+import marquez.service.models.LineageEvent;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+import org.jdbi.v3.sqlobject.transaction.Transaction;
+import org.postgresql.util.PGobject;
+
+/** The DAO for {@code dataset} facets. */
+public interface DatasetFacetsDao {
+  /* An {@code enum} used ... */
+  enum Type {
+    DATASET,
+    INPUT,
+    OUTPUT,
+    UNKNOWN;
+  }
+
+  /* An {@code enum} used to determine the dataset facet. */
+  enum DatasetFacet implements Facet {
+    DOCUMENTATION(Type.DATASET, "documentation") {
+      Optional<PGobject> toPgObject(LineageEvent.DatasetFacets facets) {
+        return Optional.ofNullable(facets.getDocumentation()).map(this::toPgObject);
+      }
+    },
+    SCHEMA(Type.DATASET, "schema") {
+      Optional<PGobject> toPgObject(LineageEvent.DatasetFacets facets) {
+        return Optional.ofNullable(facets.getSchema()).map(this::toPgObject);
+      }
+    },
+    DATASOURCE(Type.DATASET, "dataSource") {
+      Optional<PGobject> toPgObject(LineageEvent.DatasetFacets facets) {
+        return Optional.ofNullable(facets.getDataSource()).map(this::toPgObject);
+      }
+    },
+    DESCRIPTION(Type.DATASET, "description") {
+      Optional<PGobject> toPgObject(LineageEvent.DatasetFacets facets) {
+        return Optional.ofNullable(facets.getDescription()).map(this::toPgObject);
+      }
+    },
+    LIFECYCLE_STATE_CHANGE(Type.DATASET, "lifecycleStateChange") {
+      Optional<PGobject> toPgObject(LineageEvent.DatasetFacets facets) {
+        return Optional.ofNullable(facets.getLifecycleStateChange()).map(this::toPgObject);
+      }
+    },
+    VERSION(Type.DATASET, "version"),
+    COLUMN_LINEAGE(Type.DATASET, "columnLineage") {
+      Optional<PGobject> toPgObject(LineageEvent.DatasetFacets facets) {
+        return Optional.ofNullable(facets.getColumnLineage()).map(this::toPgObject);
+      }
+    },
+    OWNERSHIP(Type.DATASET, "ownership"),
+    DATA_QUALITY_METRICS(Type.INPUT, "dataQualityMetrics"),
+    DATA_QUALITY_ASSERTIONS(Type.INPUT, "dataQualityAssertions"),
+    OUTPUT_STATISTICS(Type.OUTPUT, "outputStatistics");
```

Comment:
Discussed briefly the other day, but these enums should be unnecessary, especially the facets that aren't mapped directly to Java fields - e.g., `outputStatistics`, `dataQualityMetrics`, etc.). Your call to `datasetFacets.getAdditionalFacets()` below should fetch all facets that haven't been mapped to a field in the `DatasetFacets` class. But since you have to convert everything to JSON anyway, why not convert the whole `DatasetFacets` class to a generic JsonNode object? You could do something like
```java
    JsonNode jsonNode = Utils.getMapper().valueToTree(facets);
    StreamSupport.stream(Spliterators.spliteratorUnknownSize(jsonNode.fieldNames(), Spliterator.DISTINCT),
    false)
        .forEach(fieldName -> {
          storeFacet(fieldName, toPgObject(jsonNode.get(fieldName)));
        });
```
You'll be able to support all currently known facets as well as new facets and custom facets without ever having to update the server code.

Reply:
I got rid of enums in `RunFacetsDao` and `JobFacetsDao`. 
We still need it in a reduced form in `DatasetFacetsDao` to extract facet type (`DATASET`, `INPUT`, `OUTPUT` or `UNKNOWN`).